### DISCUSSION TITLE
Where to start to develop applications like win-vind?

### DISCUSSION THREAD

**Discussion by Positron010** [pit-ray/win-vind#33] [Development] [answered]
Hi,
I really like `win-vind`! 
`win-vind` seems to be written in `C++`. Could you kindly recommend a starting place and resources to learn how to develop Windows applications like `win-vind`? Also, why did you choose `C++` over `C#`?

Thanks

**Comment by pit-ray** [comment, marked as answer]
Thank you again.

> why did you choose C++ over C#?  

win-vind handles low-level keystrokes before they are passed on to other applications, so the load needs to be minimized so that there is no delay to other applications. C# and Rust are easy to maintain, but they have a bit of overhead since the Windows API (Win32) is written in C and needs to go through DLLs and wrappers. In addition, some low-level minor functions are not supported. Therefore, it is written in C++, which is compatible with C.

> Could you kindly recommend a starting place and resources to learn how to develop Windows applications like win-vind?  

win-vind uses C++ and Win32 as a rare case today. If you want to create a general GUI application, you should use a multi-platform framework in another more maintainable language. 

The following assumes that you are developing in Win32 and C++. 

I recommend understanding the message loop to start Win32. Since Win32 has been around for a long time, there are many examples in various threads and blogs, so if you search for what you want to do, you can usually find it. Since it takes a lot of time to understand every corner of the Window API, I recommend reading the concepts or remarks of the used functions each time with the API reference

For C++, I recommend Effective C++ and Effective Modern C++, which are very practical. (But they are not introductory books).

Thanks.

**Reply by Positron010** [reply]
Thanks for replying.
